README
# Quick and Dirty functions for Wildlife simulations.

This package began life as a collection of little functions to estimate occupancy or abundance or survival directly in `R`, without having to export the data to `PRESENCE` or `MARK` or `EstimateS`, put together for a simulations workshop.

Recently we have started to use it in basic study design and data analysis workshops, where we need Quick and Dirty ways to do both MLE and Bayesian analysis for example data sets, without having to teach people to use the industry-standard software.

So we want to have data in `R` in ordinary data frames, and to use `R`'s formula interface and "twiddle" notation to define models, as in `lm`, `glm` and the `secr` package. Bayesian versions should produce output with the same look and feel as the `BEST` package. We also want it to work on all platforms: Windows, Linux, and Mac.


NEWS
CHANGES in 0.3.2 (2022-11-15)

  * Ngumbang as maintainer.

CHANGES in 0.3.1 (2022-06-19)

  * Added function `predictAvg` and data set `toves`.

  * Updated URLs in help pages.

  * Changed if(class(.) == ...) to if(inherits(...))

  * Added additional information to the help page for `BoccSS`.

  * Fixed bug in `Bnormal` and `Bnormal2`: object returned has number of chains as attribute.

  * Fixed bug in input data checks for `occ2sps`.

  * Fixed: in data set `railSims`, `logArea` is now a vector, not a 1-column matrix; values unchanged.

CHANGES in 0.3.0 (2020-06-10)

  * Added package `mcmcOutput` to Depends; all functions handling MCMC chains removed: `acfPlot`, `crosscorrPlot`, `densityFolded`, `densityPlot`, `diagPlot`, `postPriorOverlap`, `tracePlot`; functions producing MCMC chains return an `mcmcOutput` object.

  * Class `Bwiqid` deprecated; `as.Bwiqid` warns and returns an `mcmcOutput` object.

  * `Bsecr0` now returns MCMC chains for activity centres; function `plotACs` added to plot these.

  * Fixed: in data set `railSims`, `logArea` is now a vector, not a 1-column matrix; values unchanged.

  * Functions with built-in `.Time` variable get `.Time2` and `.Time3` variables; no need now for I(.Time^2).

  * Intercepts are now labeled "Intrcpt" instead of "(Intercept)", so no need for backticks.

  * Fixed: issues running examples and tests when Suggests packages are not available.

  * Spelling checked with en-GB dictionary.

CHANGES in 0.2.3 (2019-12-31)

  * Added `secrFit` wrapper for `secr::secr.fit` with `fastproximity = FALSE`

  * References to "sample" replaced with "draws" in all MCMC-related locations.

  * Fixed: bug in `dt2`; also error in argument names for `*t2` functions;
    new `*t3` functions added.

  * Fixed: bug in `occ2sps` when Hessian not invertible.

  * Converted all data files to .RData instead of R. Needed Depends: R (>= 2.10).

  * `diagPlot` revamped; accepts a range of classes, not just `Bwiqid`; additional arguments.

CHANGES in 0.2.2

  * Fixed incorrect handling of class `bugs` objects in `diagPlot` and `as.Bwiqid`.

  * `as.Bwiqid` revamped; makes more effort to retain original Rhat and n.eff; parameter
    names conversion changed: a1[1,2] now becomes a1.1.2. instead of a11.2; original
    names retained as attribute `BUGSnames`.

  * Changed `diagPlot` display for integers to dynamite plot instead of histogram;
    separate means for each chain in `diagPlot` density plots.

  * Added `crosscorrPlot` function for a range of classes.

CHANGES in 0.2.1

  * `diagPlot` no longer uses its own colour palette.

  * `window.Bwiqid` now adds new diagnostics for Rhat and n.eff.

  * Rhat calculated with `simpleRhat`, n.eff with `safeNeff` everywhere.

CHANGES in 0.2.0

  * Continuous covariates now standardised to mean=0, SD=1 (instead of SD=0.5).
    Default priors for `BoccSS` and `BsurvCJS` changed to match this.

  * Added `standardize` and `standardize2match` functions.

  * Added `simpleRhat` and `getMCerror` functions; used by `as.Bwiqid`.

  * `diagPlot` gains a `howMany` argument to plot less than the whole chain.

  * `as.Bwiqid.data.frame` gains arguments `n.chains`, `Rhat` and `n.eff` and is
    included in the help page.

  * If the variance-covariance matrix cannot be calculated, MLE functions
    now return the log(likelihood) but number of parameters, df = NA.

  * `AICc` gains a `df` argument which overrides the df value in the model
    object, so an AICc value can be got even if the covariance matrix cannot be
    calculated.

  * Added `*beta2` and `*gamma2`, wrappers for `*beta` and `*gamma` with mean and sd
    parameters, and `*beta3` with mode and concentration.

  * Added `getBetaPar` and `getGammaPar` to convert mean/sd parameters to shape/rate.

CHANGES in 0.1.5 (2018-04-20)

  * Fixed check for installation of package `shiny` in `showShinyApp`.

CHANGES in 0.1.4 (2018-04-20)

  * new function `densityFolded`, used in `plotPost` and `plot.Bwiqid` when `showCurve=TRUE`.

  * `summary.Bwiqid` method added, prints background and returns values invisibly.

  * `diagPlot` now displays `n.eff`

  * fix: in `as.Bwiqid.mcmc.list`, call to `gelman.diag` wrapped in `try`.

  * fix: histogram breaks in `plotPost` are equidistant (as always intended).

  * fix: the first argument to `BoccSS0` can be a matrix or data frame with a detection history.

  * fix: `occSSrn0` now accepts scalar `n`, as documented.

CHANGES in 0.1.3 (2017-11-18)

  * Calculation of SEs of fitted values now uses less memory, works for large data sets.

  * Functions more robust to underflow/overflow: `closedCap*`, `occ2sps`, `occSS*`

  * Changed parameterization in `secr0` from lam0 to g0; better default priors.

  * Bug fix in `BoccSS0`: burnin iterations are now discarded.

CHANGES in 0.1.2 (2017-09-10)

  * Added `diagPlot` function and improved other diagnostic plots.

  * Added function `Bpoisson`; `Bbinom` renamed `Bbinomial`

CHANGES in 0.1.1 (2017-06-09)

  * Added `coef`, `vcov` and `predict` methods for `wiqid` objects.

  * `survCJS` gains an `interval` argument to deal with time intervals of unequal length.

  * dots argument added to MLE functions to pass parameters to `nlm`.

  * Improved coding of several functions to prevent under/overflow issues.


R/utils.R

# This file contains utilities used in several places in the code
#   and NOT exported:

# modelMatrix : wrapper for model.matrix, changes "(Intercept)" to "Intrcpt"
# getVar0, getFittedVar : get variance for fitted values
# getScaling, doScaling, scaleToMatch : functions to deal with scaling
# signifish : an alternative to signif (added 10-02-2015)
# fixCI : Calculate critical values for CI.
# getMARKci : Calculate MARK-style confidence intervals for N
# stdModel : Regularize a list of formulae, ensuring it is a named list of one-sided formulae.
# stddata : Convert a data frame of site and survey data into a list and standardise
# selectCovars : Pull the covars needed for a model matrix into a specific data frame
# matchStart : has its own file

# fixNames : removed, use 'make.names(., unique=TRUE) instead (2019-04-17)
# AICtable moved to file AICc.R
# logSumExp etc are now in file UnderOverflow.R
# Functions to convert parameters of distributions (eg mean and sd to shape and rate)
#   are in converters.R
# Variants of the t-distribution are in TDist.R
# ...............................................................................

# Produces a  model matrix but the intercept column is named "Intrcpt" without
#   parentheses
modelMatrix <- function(formula, data, ...) {
  mm <- model.matrix(formula, data, ...)
  colnames(mm)[colnames(mm) == "(Intercept)"] <- "Intrcpt"
  return(mm)
}

# ...............................................................................

# Functions to calculate the variance of fitted values from model matrix and var-covar matrix.
# added 2017-10-16
# Output of getFittedVar is equivalent to
#   diag(MM %*% varcov %*% t(MM))
# but does not require calculation of the full matrix, which can be huge.

getVar0 <- function(x, vcv)
  x %*% vcv %*% x
# x : one row of a model matrix
# vcv : variance-covariance matrix

getFittedVar <- function(MM, vcv)
  apply(MM, 1, getVar0, vcv=vcv)
# MM : a model matrix
# vcv : variance-covariance matrix

# ...............................................................................

# Functions to deal with scaling, can be used with s/lapply

getScaling <- function(x, scaleBy)
  c(mean(x, na.rm=TRUE), sd(x, na.rm=TRUE) / scaleBy)

doScaling <- function(x, scaleBy)
  if(is.numeric(x)) (x - mean(x, na.rm=TRUE)) / sd(x, na.rm=TRUE) * scaleBy else x

# This takes a whole data frame
scaleToMatch <- function(target, scaling) {
  for(i in seq_along(target)) {
    if(is.numeric(target[[i]])) {
      pos <- match(names(target)[i], names(scaling))
      if(!is.na(pos)) {
        sc <- scaling[[pos]]
        target[[i]] <- (target[[i]] - sc[1]) / sc[2]
      }
    }
  }
  return(target)
}

# ...............................................................................

# A more sensible version of signif
signifish <- function(x, digits=3)
  ifelse(x < 10^digits, signif(x, digits=digits), round(x))

# ...............................................................................

# Deal with confidence interval specification:
fixCI <- function(ci) {
  if(ci > 1 | ci < 0.5)
    stop("ci must be between 0.5 and 1")
  alf <- (1 - ci[1]) / 2
  return(qnorm(c(alf, 1 - alf)))
}
# .....................................................................

# Function to calculate the MARK-style confidence intervals for N
# See help for Closed Captures

getMARKci <- function(beta, SE.beta, ci) {
  f0.hat <- exp(beta)
  crit <- qnorm((1 - ci[1]) / 2, lower.tail=FALSE)
  C <- exp(crit * sqrt(log(1 + SE.beta^2))) # See the Burnham et al reference, p212!
  return(c(f0.hat, f0.hat/C, f0.hat*C))
}
# .........................................................................

## Regularize a list of formulae, ensuring it is a named list of one-sided formulae.
# New version:
stdModel <- function (model1, defaultModel) {
  if(is.null(model1))
    return(defaultModel)
  if(inherits(model1, "formula"))
    model1 <- list(model1)
  if(!is.list(model1))
    stop("The 'model' argument must be a formula or a list of formulae.")
  LHS <- function (form) {
      trms <- as.character (form)
      if (length(trms)==2) '' else trms[2]
  }
  RHS <- function (form) {
      trms <- as.character (form)
      if (length(trms)==3) as.formula(paste(trms[c(1,3)], collapse=" ")) else form
  }
  lhs <- sapply(model1, LHS)
  temp <- lapply(model1, RHS)
  if (is.null(names(model1))) {
    names(temp) <- lhs
  } else {
    names(temp) <- ifelse(names(model1) == '', lhs, names(model1))
  }
  newModel <- replace (defaultModel, names(temp), temp)
  return(newModel)
}
# .............................................................................

## Convert a data frame of site and survey data into a list
# ** Site covars will each have a single column in the data frame,
# ** survey covars will have a column for each survey occasion, and
# column names end with the number of the occasion, eg, temperature
# will be in columns named "temp1", "temp2", etc.

stddata <- function(df, nocc=NULL, scaleBy=1)  {
  if (is.null(df))
    return(NULL)
  if(!is.data.frame(df))
    stop("The 'data' argument must be a data frame.")
  dataList <- as.list(df)
  ## Group variables spread over > 1 column into a single vector
  if (!is.null(nocc)) {
    nocc <- sort(nocc, decreasing=TRUE) # start with biggest
    for (this.nocc in nocc)  {
      # look for names ending with number of occasions
      nam <- names(df)
      clue <- paste0(this.nocc, "$", collapse="")
      clueDo <- grep(clue, nam)
      if(length(clueDo) > 0) {
        for(i in clueDo) {
          # get stem, generate set of names
          stem <- sub(clue, "", nam[i])
          subnames <- paste0(stem, 1:this.nocc)
          subtable <- df[, subnames]
          # check that there's a column for each occasion
          if(ncol(subtable) != this.nocc)
            stop("Survey covariates must have 1 column for each survey occasion.")
          # check that all have same class
          classes <- sapply(subtable, class)
          if(length(unique(classes)) != 1)
            stop("All columns of the survey covariates must have the same class (all factor or all numeric).")
          # remove original columns from the list:
          dataList <- replace(dataList, subnames, NULL)
          # convert to a matrix, then a vector;
          #   fortunately this also converts factors to character
          tmp <- as.vector(as.matrix(subtable))
          if(is.character(tmp))
            tmp <- as.factor(tmp) # convert to factor AFTER combining columns
          dataList <- c(dataList, list(tmp))
          names(dataList)[length(dataList)] <- stem
        }
      }
    }
  }
  ## Standardize numeric variables to mean = 0 and sd = scaleBy
  if (!is.null(scaleBy)) {
    doScale <- function(x) {
      if (is.numeric(x))
        x <- standardize(x) * scaleBy
      return(x)
    }
    dataList <- lapply(dataList, doScale)
  }
  return(dataList)
} # ...........................................................................

# Pull the covars needed for a model matrix into a specific data frame
selectCovars <- function(formula, dataList, minrows)  {
  wanted <- rownames(attr(terms(formula), "factors"))
  found <- wanted %in% names(dataList)
  wanted <- wanted[found]
  if (length(wanted) > 0)  {
    df <- as.data.frame(dataList[wanted])
    df <- cbind(df, .dummy = rep(NA, minrows))
  } else {
    df <- data.frame(.dummy = rep(NA, minrows))
  }
  stopifnot(nrow(df) %% minrows == 0)
  return(df)
} # ........................................................




